platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformEdgeCasesIntegrationTest.groovy

```diff
@@ -0,0 +1,902 @@
+        executer.expectDeprecationWarning("There are multiple distinct artifact transformation chains of the same length that would satisfy this request. This behavior has been deprecated. This will fail with an error in Gradle 9.0. ")
+        succeeds "resolve"
+
+        then:
+        output.count("Transforming") == 2
+        output.count("Transforming lib.jar to lib.jar.txt") == 1
+        output.count("Transforming test-1.3.jar to test-1.3.jar.txt") == 1
+
+        when: "if an additional attribute is present on both chains, then we produce distinct, non-mutually compatible variants, and fail with ambiguity"
+        fails 'resolve', '-PextraAttributeA', '-PextraAttributeB'
+
+        then:
+        failureCauseContains('Found multiple transformation chains')
+    }
+
+    def "when A -> C and B -> C both produce identical attributes, adding an additional attribute to either removes the latent ambiguity and causes the other to be selected as the better match"() {
+        given:
+        setupDisambiguationTest()
+
+        when: "if only transform A produces an extra attribute, then it produces a final variant that is farther from the request because of it, so B is the winner"
+        succeeds 'resolve', '-PextraAttributeA'
+
+        then:
+        output.count("Transforming") == 2
+        output.count("Transforming lib.jar to lib.jar.txt") == 1
+        output.count("Transforming test-1.3.jar to test-1.3.jar.txt") == 1
+
+        when: "and vice-versa with B"
+        succeeds 'resolve', '-PextraAttributeB'
+
+        then:
+        output.count("Transforming") == 1
+        output.count("Transforming main to main.txt") == 1
+    }
+
+    def "when A -> C and B -> C differ by a single discriminating attribute, adding that additional attribute resolves the ambiguity"() {
+        given:
+        setupDisambiguationTest()
+
+        when:
+        succeeds 'resolve', '-PextraAttributeA', '-PextraAttributeB', '-PextraRequest=value2'
+
+        then:
+        output.count("Transforming") == 2
+        output.count("Transforming lib.jar to lib.jar.txt") == 1
+        output.count("Transforming test-1.3.jar to test-1.3.jar.txt") == 1
+    }
+
+    private void setupDisambiguationTest() {
+        def m1 = mavenRepo.module("test", "test", "1.3").publish()
+        m1.artifactFile.text = "1234"
+
+        createDirs("lib", "app")
+        settingsFile << """
+            rootProject.name = 'root'
+            include 'lib'
+            include 'app'
+        """
+
+        file('lib/src/main/java/test/MyClass.java') << """
+package test;
+
+public class MyClass {
+    public static void main(String[] args) {
+        System.out.println("Hello world!");
+    }
+}
+"""
+
+        buildFile << """
+def artifactType = Attribute.of('artifactType', String)
+def extraAttribute = Attribute.of('extra', String)
+
+allprojects {
+    repositories {
+        maven { url "${mavenRepo.uri}" }
+    }
+}
+project(':lib') {
```

Comment:
Is there a reason we are using this `project` syntax instead of a custom build file. AFAIK this is not IP compatible so we should not add new tests using this syntax.

Reply:
Just copy-paste.  I'll remove this.